### ISSUE TITLE
The `no-magic-numbers` rule should ignore function calls 

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#4236]
Usually when one number is used with a function the intent is clear and I wouldn't consider that a magic number. Currently this rule is unusable for me as it gives too many false positives.

In my tests I use a `t.plan()` method to plan how many assertion the test will have. I have a lot of the following, so my tests gets an endless amount of `no-magic-number` violations:

```js
test('foo', function (t) {
  t.plan(2);
  ...
});
```

-

```
~/dev
❯ eslint --version
v1.7.3
                                                                                                                                               
~/dev
❯ echo 't.plan(2)' | eslint --stdin --rule='no-magic-numbers: [2, {ignore: []}]'

<text>
  1:8  error  No magic number: 2  no-magic-numbers

✖ 1 problem (1 error, 0 warnings)
```

-

The rule should either ignore function calls by default or have an option for doing so.

**Comment by nzakas** [comment]
Can you explain why you think function calls should be an exception?

**Comment by scottohara** [comment]
It is an interesting question: when should a raw integer be considered 'magic'?
Always? Or only in certain circumstances?

(Just to be clear - I'm only adding commentary to the discussion here; I'm not arguing one way or another).

The docs have good examples of what I personally would typically consider as "magic numbers".  Possibly one of the reasons I consider them magic is that they're used in a _calculation_ of some sort, i.e.

```
/*eslint no-magic-numbers: 2*/

var dutyFreePrice = 100,
    finalPrice = dutyFreePrice + (dutyFreePrice * 0.25); /*error No magic number: 0.25*/
```

versus

```
/*eslint no-magic-numbers: 2*/

var TAX_PERCENTAGE = 0.25;

var dutyFreePrice = 100,
    finalPrice = dutyFreePrice + (dutyFreePrice * TAX_PERCENTAGE);
```

But is _"used in a calculation"_ a reliable discriminator of whether a number is magic or not?  I don't know (and could we even practically test that?).

Some examples of numbers that I would **not** typically consider as "magic" would be:

`let firstChar = someString.charAt(0);    // getting Nth char from a string`
`let restOfString = someString.substring(1)  // getting a substring`
`myArray.splice(index, 1);  // removing a entry from an array`
`if (myArray.length > 0)   // testing if array is empty`
`myArray.reduce(fn, 0)   // array reduce, with starting memo of zero`
`myArray.sort((a, b) => a < b ? -1 : a > b ? 1 : 0);  // custom array sort fn`
`window.setTimeout(fn, 0) // detach from call stack`
`if (someString.indexOf(someOtherString) !== -1)  // old-style string contains (before ES2015 String.prototype.includes())`

Of course, configuring the rule as `no-magic-numbers: [2, {ignore: [-1, 0, 1]}]` would allow all of the above examples to pass, and implies that these whitelisted are somehow "special", yet not "magic".

If a developer's code needs to extract the 3rd character from a string (`someString.charAt(2)`);  the developer must decide whether to whitelist the value `2` as being "not magic", or they need to (as the docs suggest) refactor it to a constant (`const INDEX_POSITION_THREE = 2;  let thirdChar = myString.charAt(INDEX_POSITION_THREE);`).

I can certainly see merit in @sindresorhus 's comment that a lone integer, passed as a argument to a function isn't usually considered magic. In my own code, I also found too many false positives (or a need to whitelist too many 'special' numbers) for this rule to be useful for me.

I can also see, however, that it might be a slippery slope if single-int function params were allowed as an exception to the rule....because then where do you stop?  Having a blanket "any raw numbers (except those on a whitelist) are considered magic" is obviously far easier to maintain.

The key here is finding a good definition of "magic number" that most people can agree on.

**Comment by nzakas** [comment]
As far as I know, there is no definition that states function calls should be ignored. We try very hard not to make rules that change the meaning of well-known terms, so what I'm looking for is some documentation that shows omitting function calls is an acceptable practice with magic numbers.

**Comment by sindresorhus** [comment]
@nzakas There's no definite answer of what a magic number is either. Magic numbers by most explanations are numbers with unclear intent. With `t.plan(2);` the intent is clear. I could have done `var PLAN_COUNT = 2; t.plan(PLAN_COUNT);` but it wouldn't have made it any clearer. As for documentation, I think @scottohara documented it pretty clearly above. I don't think you're going to find any better documentation elsewhere. At least I couldn't. I like the idea of this rule, but right now it's not very practical. We should be able to find a middle ground between the current situation and too much slippery slope. I think a function call with one argument magic number would be a good exception as the intent would always be the function name.

**Comment by nzakas** [comment]
> I think a function call with one argument magic number would be a good exception as the intent would always be the function name.

I don't think "one argument is always clear" is necessarily a true statement. Our problem is that we cannot definitively know if intent is clear or not. For instance, I wouldn't know what `t.plan(2)` meant if I saw it without your description above, so to me is a magic number vs. `t.plan(EXPECTED_ASSERTS)`.

The other issue we have is that strangely specific exception cases are often misunderstood by users and opens us up to unending requests for more strangely specific exceptions. 

I'm open to adding an exception that excludes all function calls, but I think that's about as far as we can go here. 

And remember, you can always create a rule that does exactly what you want, as well.

**Comment by nzakas** [comment]
Seems like we couldn't resolve this, and no further interest or discussion occurred in the past few months, so closing.
